Toward removing main.py as a dependence

powerjg opened the pull request:
I am working toward enabling a more modular version of gem5 where we can link to python modules instead of using the gem5 binary with the marshalled python code. This is one of the many hurdles to overcome: specifically that main.py populates m5.options. This PR is the first step toward decoupling the stdlib from main.py and m5.options. Generally, this is cleaning up code and there should be no functional changes.

powerjg committed:
stdlib: Remove deprecated checkpoint param from simulator

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python: split _instantiate into multiple functions

The near-term need to break the dependence of _instantiate on options.
Further in the future, this will allow others to more easily reimplement main.py

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python: Make gather_citations match other output calls

Instead of using m5.options to get the outdir in gather_citations,
this change makes gather_citations mathc the other output functions
and passes the directory as a parameter

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python: Allow instantiate to work without options

This change updates the new instantiate sub-functions to take the output
directory as an optional parameter. They also take the file locations
for the config and dot files as parameters. This will allow users
to override the output directory without depending on main.py and options.py

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python: Expose the simulate module in m5

Currently, because we use from m5.simulate import * in the
m5.__init__ file the m5.simulate module is impossible to import.
This change exposes that module as _simulate_module.

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
stdlib: Implement instantiate steps in Simulator

Instead of calling m5.instantiate() this change moves (copies) the
implementatino of m5.instantiate() into the Simulator _instantiate
function. It also allows the simulator to set the outdir which
is a step to decoupling the Simulator class from m5.main and
m5.options.

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
stdlib: Fix boards overriding _pre_instantiate

The _pre_instantiate function in AbstractBoard returns a Root object.
The Arm and RISC-V boards did not correctly return the root object.

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

BobbyRBruce commented on src/python/m5/simulate.py:
```diff
@@ -103,32 +90,75 @@ def instantiate(ckpt_dir=None):
+        from m5 import options
+
+        outdir = options.outdir
+    if dump_config is None:
```
Any particular reason we couldn't take this opportunity to rename this something like `ini_config`? Just to be a bit more explicit? From my understanding we dump 3 different versions of the configs. This is just one of them

powerjg replied:
Done.

BobbyRBruce commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -101,9 +101,9 @@ def __init__(
             ]
         ] = None,
         expected_execution_order: Optional[List[ExitEvent]] = None,
-        checkpoint_path: Optional[Path] = None,
         max_ticks: Optional[int] = m5.MaxTick,
         id: Optional[int] = None,
+        outdir: Optional[str | Path] = None,
     ) -> None:
```
I take it the intention here is that the user specifies the checkpoint path when you take the checkpoint. If so, I'd be careful here: I believes some exit events/hyper calls expect to take a checkpoint with the path already set. I believe this why this is in the constructor: It  enables someone to set up a default checkpoint path for these exit event handlers to dump stuff in.

powerjg replied:
This is what's causing the CI tests to fail. I'm confused as to why we're dependent on this if it's deprecated.

BobbyRBruce replied:
I think while this was in a deprecated state, we started using it for Multisim. In Multisim, you can't declare anything to be set up after instantiation for a given gem5 child process system, -- you have to do everything beforehand. So, this became a necessity. We should have taken it out of a deprecated state (or figured out a way around this problem we require it for), but I guess we lost track of it.

powerjg replied:
I updated this a bit. The checkpoint_path seems to be part of the board now. I've added the needed APIs to expose it.

BobbyRBruce reviewed (commented):
In addition to my comments, I'd also strongly suggest running some MultiSim runs to verify this doesn't break anything. I can't remember exactly how it all fits together, but I remember having to hack quite a lot of the `m5.outdir` and `main.py` code to get all the various gem5 subprocesses to output the right files to the right directories. Given you're moving this code around a bit, it'd be worthwhile checking that all still works. There are some examples in "configs/example/gem5_librabry".

powerjg committed:
stdlib: Add API to get checkpoint dir

This will enable us to remove the checkpoint dir as a parameter
to the simulator.

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>
